File: src-tauri/src/commands/web_proxy_credentials.rs
use aes_gcm::{
    aead::{Aead, KeyInit, Payload},
    Aes256Gcm, Nonce,
};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use hkdf::Hkdf;
use rand_core::OsRng;
use serde::{Deserialize, Serialize};
use sha2::Sha256;
use std::time::Duration;
use url::Url;
use x25519_dalek::{PublicKey, StaticSecret};
use zeroize::Zeroize;

const CREDENTIAL_PATH: &str = "/_jumpserver/web-sessions/";
const CREDENTIAL_KDF_INFO: &[u8] = b"jumpserver-web-autofill-v1";
const X25519_SPKI_PREFIX: &[u8] = &[
    0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x6e, 0x03, 0x21, 0x00,
];

#[derive(Debug, Deserialize)]
struct CreateCredentialResponse {
    id: Option<String>,
    access_token: Option<String>,
    target_url: String,
    origin: String,
    autofill_available: bool,
    username_selector: Option<String>,
    password_selector: Option<String>,
    submit_selector: Option<String>,
    server_public_key: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ReleasedCredentialResponse {
    nonce: String,
    ciphertext: String,
}

#[derive(Debug, Deserialize, Zeroize)]
#[zeroize(drop)]
pub(crate) struct WebCredentials {
    pub(crate) username: String,
    pub(crate) password: String,
}

pub(crate) struct WebCredentialSession {
    id: String,
    access_token: String,
    endpoint: Url,
    origin: String,
    username_selector: String,
    password_selector: String,
    submit_selector: String,
    server_public_key: [u8; 32],
    private_key: StaticSecret,
}

impl Drop for WebCredentialSession {
    fn drop(&mut self) {
        self.access_token.zeroize();
    }
}

#[derive(Serialize)]
struct CreateCredentialRequest<'a> {
    token_id: &'a str,
    token_value: &'a str,
    client_public_key: String,
}

impl WebCredentialSession {
    pub(crate) fn origin(&self) -> &str {
        &self.origin
    }

    pub(crate) fn selectors(&self) -> (&str, &str, &str) {
        (
            &self.username_selector,
            &self.password_selector,
            &self.submit_selector,
        )
    }

    pub(crate) async fn release(&mut self, current_url: &Url) -> Result<WebCredentials, String> {
        if normalized_origin(current_url)? != self.origin {
            return Err("页面 origin 与代填会话不匹配".to_string());
        }
        let url = self
            .endpoint
            .join(&format!("{}/credentials", self.id))
            .map_err(|error| error.to_string())?;
        let client = credential_client()?;
        let response = client
            .post(url)
            .bearer_auth(&self.access_token)
            .json(&serde_json::json!({ "origin": self.origin }))
            .send()
            .await
            .map_err(|error| format!("领取 Web 凭据失败: {error}"))?;
        self.access_token.zeroize();
        if !response.status().is_success() {
            return Err(format!("领取 Web 凭据失败: HTTP {}", response.status()));
        }
        let released: ReleasedCredentialResponse = response
            .json()
            .await
            .map_err(|error| format!("解析 Web 凭据失败: {error}"))?;
        self.decrypt(released)
    }

    fn decrypt(&self, released: ReleasedCredentialResponse) -> Result<WebCredentials, String> {
        let server_key = PublicKey::from(self.server_public_key);
        let mut shared_secret = self.private_key.diffie_hellman(&server_key).to_bytes();
        let hkdf = Hkdf::<Sha256>::new(None, &shared_secret);
        let mut key = [0u8; 32];
        hkdf.expand(CREDENTIAL_KDF_INFO, &mut key)
            .map_err(|_| "派生 Web 凭据密钥失败".to_string())?;
        shared_secret.zeroize();

        let nonce = BASE64
            .decode(released.nonce)
            .map_err(|_| "Web 凭据 nonce 无效".to_string())?;
        let ciphertext = BASE64
            .decode(released.ciphertext)
            .map_err(|_| "Web 凭据密文无效".to_string())?;
        if nonce.len() != 12 || ciphertext.len() <= 16 {
            key.zeroize();
            return Err("Web 凭据密文格式无效".to_string());
        }
        let cipher =
            Aes256Gcm::new_from_slice(&key).map_err(|_| "初始化 Web 凭据解密器失败".to_string())?;
        let aad = format!("{}\n{}", self.id, self.origin);
        let mut plaintext = cipher
            .decrypt(
                Nonce::from_slice(&nonce),
                Payload {
                    msg: &ciphertext,
                    aad: aad.as_bytes(),
                },
            )
            .map_err(|_| "Web 凭据校验失败".to_string())?;
        key.zeroize();
        let result = serde_json::from_slice(&plaintext).map_err(|_| "Web 凭据内容无效".to_string());
        plaintext.zeroize();
        result
    }
}

pub(crate) async fn create_credential_session(
    proxy_url: &Url,
    target_url: &Url,
    token_id: &str,
    token_value: &str,
) -> Result<Option<WebCredentialSession>, String> {
    if token_id.is_empty() || token_value.is_empty() {
        return Ok(None);
    }
    let private_key = StaticSecret::random_from_rng(OsRng);
    let public_key = PublicKey::from(&private_key);
    let mut spki = Vec::with_capacity(X25519_SPKI_PREFIX.len() + public_key.as_bytes().len());
    spki.extend_from_slice(X25519_SPKI_PREFIX);
    spki.extend_from_slice(public_key.as_bytes());
    let endpoint = proxy_url
        .join(CREDENTIAL_PATH)
        .map_err(|error| error.to_string())?;
    let client = credential_client()?;
    let response = client
        .post(endpoint.clone())
        .json(&CreateCredentialRequest {
            token_id,
            token_value,
            client_public_key: BASE64.encode(spki),
        })
        .send()
        .await
        .map_err(|error| format!("创建 Web 代填会话失败: {error}"))?;
    if !response.status().is_success() {
        return Err(format!("创建 Web 代填会话失败: HTTP {}", response.status()));
    }
    let data: CreateCredentialResponse = response
        .json()
        .await
        .map_err(|error| format!("解析 Web 代填会话失败: {error}"))?;
    let response_target =
        Url::parse(&data.target_url).map_err(|_| "Koko 返回了无效 Website 地址".to_string())?;
    if normalized_origin(&response_target)? != normalized_origin(target_url)?
        || normalized_origin(&response_target)? != data.origin
    {
        return Err("Koko 返回的 Website origin 不匹配".to_string());
    }
    if !data.autofill_available {
        return Ok(None);
    }

    let username_selector = data.username_selector.unwrap_or_default();
    let password_selector = required_selector(data.password_selector, "密码")?;
    let submit_selector = required_selector(data.submit_selector, "提交")?;
    if !username_selector.is_empty() {
        validate_selector(&username_selector)?;
    }
    let server_public_key = parse_spki(&data.server_public_key.unwrap_or_default())?;
    Ok(Some(WebCredentialSession {
        id: required(data.id, "代填会话 ID")?,
        access_token: required(data.access_token, "代填访问令牌")?,
        endpoint,
        origin: data.origin,
        username_selector,
        password_selector,
        submit_selector,
        server_public_key,
        private_key,
    }))
}

fn credential_client() -> Result<reqwest::Client, String> {
    reqwest::Client::builder()
        .timeout(Duration::from_secs(15))
        .no_proxy()
        .build()
        .map_err(|error| error.to_string())
}

fn required(value: Option<String>, field: &str) -> Result<String, String> {
    value
        .filter(|item| !item.is_empty())
        .ok_or_else(|| format!("Koko 返回的{field}为空"))
}

fn required_selector(value: Option<String>, label: &str) -> Result<String, String> {
    let selector = required(value, &format!("{label}元素配置"))?;
    validate_selector(&selector)?;
    Ok(selector)
}

fn validate_selector(selector: &str) -> Result<(), String> {
    if selector.is_empty() || selector.len() > 1024 {
        return Err("Website 代填元素配置无效".to_string());
    }
    let (kind, value) = selector
        .split_once('=')
        .ok_or_else(|| "Website 代填元素配置无效".to_string())?;
    if value.trim().is_empty()
        || !matches!(
            kind.trim().to_ascii_lowercase().as_str(),
            "name" | "id" | "type" | "class_name" | "css" | "css_selector" | "xpath"
        )
    {
        return Err("Website 代填元素配置无效".to_string());
    }
    Ok(())
}

fn parse_spki(value: &str) -> Result<[u8; 32], String> {
    let der = BASE64
        .decode(value)
        .map_err(|_| "Koko Web 公钥无效".to_string())?;
    if der.len() != X25519_SPKI_PREFIX.len() + 32 || !der.starts_with(X25519_SPKI_PREFIX) {
        return Err("Koko Web 公钥格式无效".to_string());
    }
    der[X25519_SPKI_PREFIX.len()..]
        .try_into()
        .map_err(|_| "Koko Web 公钥格式无效".to_string())
}

pub(crate) fn normalized_origin(url: &Url) -> Result<String, String> {
    let host = url
        .host_str()
        .ok_or_else(|| "Website origin 无效".to_string())?
        .to_ascii_lowercase();
    let port = url
        .port_or_known_default()
        .ok_or_else(|| "Website origin 无效".to_string())?;
    let default_port =
        (url.scheme() == "http" && port == 80) || (url.scheme() == "https" && port == 443);
    let authority = if default_port {
        host
    } else {
        format!("{host}:{port}")
    };
    Ok(format!(
        "{}://{}",
        url.scheme().to_ascii_lowercase(),
        authority
    ))
}

#[cfg(test)]
mod tests {
    use super::{
        parse_spki, validate_selector, ReleasedCredentialResponse, WebCredentialSession,
        CREDENTIAL_KDF_INFO, CREDENTIAL_PATH, X25519_SPKI_PREFIX,
    };
    use aes_gcm::{
        aead::{Aead, KeyInit, Payload},
        Aes256Gcm, Nonce,
    };
    use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
    use hkdf::Hkdf;
    use rand_core::OsRng;
    use sha2::Sha256;
    use url::Url;
    use x25519_dalek::{PublicKey, StaticSecret};

    #[test]
    fn validates_supported_selectors_and_x25519_spki() {
        assert!(validate_selector("id=password").is_ok());
        assert!(validate_selector("css=input[type=password]").is_ok());
        assert!(validate_selector("javascript=alert(1)").is_err());
        let mut spki = X25519_SPKI_PREFIX.to_vec();
        spki.extend_from_slice(&[7u8; 32]);
        assert_eq!(parse_spki(&BASE64.encode(spki)).unwrap(), [7u8; 32]);
    }

    #[test]
    fn decrypts_koko_compatible_one_time_envelope() {
        let client_key = StaticSecret::random_from_rng(OsRng);
        let server_key = StaticSecret::random_from_rng(OsRng);
        let server_public = PublicKey::from(&server_key);
        let shared = server_key.diffie_hellman(&PublicKey::from(&client_key));
        let hkdf = Hkdf::<Sha256>::new(None, shared.as_bytes());
        let mut key = [0u8; 32];
        hkdf.expand(CREDENTIAL_KDF_INFO, &mut key).unwrap();
        let nonce = [3u8; 12];
        let plaintext = br#"{"username":"managed-user","password":"managed-password"}"#;
        let ciphertext = Aes256Gcm::new_from_slice(&key)
            .unwrap()
            .encrypt(
                Nonce::from_slice(&nonce),
                Payload {
                    msg: plaintext,
                    aad: b"session-id\nhttps://example.com",
                },
            )
            .unwrap();
        let session = WebCredentialSession {
            id: "session-id".to_string(),
            access_token: "once".to_string(),
            endpoint: Url::parse("http://127.0.0.1:5001/_jumpserver/web-sessions/").unwrap(),
            origin: "https://example.com".to_string(),
            username_selector: "id=username".to_string(),
            password_selector: "id=password".to_string(),
            submit_selector: "id=submit".to_string(),
            server_public_key: server_public.to_bytes(),
            private_key: client_key,
        };
        let credentials = session
            .decrypt(ReleasedCredentialResponse {
                nonce: BASE64.encode(nonce),
                ciphertext: BASE64.encode(ciphertext),
            })
            .unwrap();
        assert_eq!(credentials.username, "managed-user");
        assert_eq!(credentials.password, "managed-password");
        assert_eq!(
            Url::parse("http://127.0.0.1:5001")
                .unwrap()
                .join(CREDENTIAL_PATH)
                .unwrap()
                .join("session-id/credentials")
                .unwrap()
                .path(),
            "/_jumpserver/web-sessions/session-id/credentials"
        );
    }
}

File: electron/web-proxy-credentials.mjs
import { createDecipheriv, createPublicKey, diffieHellman, generateKeyPairSync, hkdfSync } from "node:crypto";
import { requestWebProxyControl } from "./web-proxy-control.mjs";

const CREDENTIAL_PATH = "/_jumpserver/web-sessions/";
const CREDENTIAL_KDF_INFO = Buffer.from("jumpserver-web-autofill-v1");
const X25519_SPKI_PREFIX = Buffer.from([0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x6e, 0x03, 0x21, 0x00]);
const SUPPORTED_SELECTOR_KINDS = new Set(["name", "id", "type", "class_name", "css", "css_selector", "xpath"]);

async function fetchWithTimeout(proxyUrl, path, options, timeout = 15_000) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeout);
  try {
    return await requestWebProxyControl(proxyUrl, path, { ...options, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

function required(value, field) {
  if (typeof value !== "string" || !value) throw new Error(`Koko 返回的${field}为空`);
  return value;
}

async function responseError(response, action) {
  const detail = (await response.text()).trim();
  return new Error(`${action}: ${detail || `HTTP ${response.status}`}`);
}

export function normalizedWebOrigin(value) {
  const url = value instanceof URL ? value : new URL(value);
  if (!["http:", "https:"].includes(url.protocol) || !url.hostname) throw new Error("Website origin 无效");
  return url.origin.toLowerCase();
}

export function validateWebSelector(selector) {
  if (typeof selector !== "string" || !selector || selector.length > 1024) {
    throw new Error("Website 代填元素配置无效");
  }
  const separator = selector.indexOf("=");
  const kind = selector.slice(0, separator).trim().toLowerCase();
  const value = selector.slice(separator + 1).trim();
  if (separator < 1 || !value || !SUPPORTED_SELECTOR_KINDS.has(kind)) {
    throw new Error("Website 代填元素配置无效");
  }
  return selector;
}

export async function createCredentialSession(proxyUrl, targetUrl, tokenId, tokenValue) {
  if (!tokenId || !tokenValue) return null;

  const { privateKey, publicKey } = generateKeyPairSync("x25519");
  const endpoint = new URL(CREDENTIAL_PATH, proxyUrl);
  let response;
  try {
    response = await fetchWithTimeout(proxyUrl, endpoint.pathname, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        token_id: tokenId,
        token_value: tokenValue,
        client_public_key: publicKey.export({ type: "spki", format: "der" }).toString("base64")
      })
    });
  } catch (error) {
    throw new Error(`创建 Web 代填会话失败: ${error}`);
  }
  if (!response.ok) throw await responseError(response, "创建 Web 代填会话失败");

  let data;
  try {
    data = await response.json();
  } catch (error) {
    throw new Error(`解析 Web 代填会话失败: ${error}`);
  }
  const responseTarget = new URL(data.target_url);
  const targetOrigin = normalizedWebOrigin(targetUrl);
  if (normalizedWebOrigin(responseTarget) !== targetOrigin || String(data.origin).toLowerCase() !== targetOrigin) {
    throw new Error("Koko 返回的 Website origin 不匹配");
  }
  if (!data.autofill_available) return null;

  const usernameSelector = data.username_selector || "";
  if (usernameSelector) validateWebSelector(usernameSelector);
  const passwordSelector = validateWebSelector(required(data.password_selector, "密码元素配置"));
  const submitSelector = validateWebSelector(required(data.submit_selector, "提交元素配置"));
  const serverPublicKey = Buffer.from(required(data.server_public_key, "Web 公钥"), "base64");
  if (
    serverPublicKey.length !== X25519_SPKI_PREFIX.length + 32 ||
    !serverPublicKey.subarray(0, X25519_SPKI_PREFIX.length).equals(X25519_SPKI_PREFIX)
  ) {
    throw new Error("Koko Web 公钥格式无效");
  }

  return {
    id: required(data.id, "代填会话 ID"),
    accessToken: required(data.access_token, "代填访问令牌"),
    endpoint,
    origin: targetOrigin,
    selectors: { username: usernameSelector, password: passwordSelector, submit: submitSelector },
    serverPublicKey,
    privateKey
  };
}

export async function releaseCredentials(session, currentUrl) {
  if (normalizedWebOrigin(currentUrl) !== session.origin) throw new Error("页面 origin 与代填会话不匹配");
  const url = new URL(`${encodeURIComponent(session.id)}/credentials`, session.endpoint);
  const accessToken = session.accessToken;
  session.accessToken = "";
  let response;
  try {
    response = await fetchWithTimeout(session.endpoint, url.pathname, {
      method: "POST",
      headers: {
        authorization: `Bearer ${accessToken}`,
        "content-type": "application/json"
      },
      body: JSON.stringify({ origin: session.origin })
    });
  } catch (error) {
    throw new Error(`领取 Web 凭据失败: ${error}`);
  }
  if (!response.ok) throw await responseError(response, "领取 Web 凭据失败");

  let released;
  try {
    released = await response.json();
  } catch (error) {
    throw new Error(`解析 Web 凭据失败: ${error}`);
  }
  const nonce = Buffer.from(String(released.nonce || ""), "base64");
  const ciphertext = Buffer.from(String(released.ciphertext || ""), "base64");
  if (nonce.length !== 12 || ciphertext.length <= 16) throw new Error("Web 凭据密文格式无效");

  const serverKey = createPublicKey({ key: session.serverPublicKey, type: "spki", format: "der" });
  const sharedSecret = diffieHellman({ privateKey: session.privateKey, publicKey: serverKey });
  const key = Buffer.from(hkdfSync("sha256", sharedSecret, Buffer.alloc(0), CREDENTIAL_KDF_INFO, 32));
  sharedSecret.fill(0);
  const tagOffset = ciphertext.length - 16;
  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAAD(Buffer.from(`${session.id}\n${session.origin}`));
  decipher.setAuthTag(ciphertext.subarray(tagOffset));
  let plaintext;
  try {
    plaintext = Buffer.concat([decipher.update(ciphertext.subarray(0, tagOffset)), decipher.final()]);
  } catch {
    throw new Error("Web 凭据校验失败");
  } finally {
    key.fill(0);
  }
  try {
    const credentials = JSON.parse(plaintext.toString("utf8"));
    if (typeof credentials.username !== "string" || typeof credentials.password !== "string") {
      throw new Error("invalid credential fields");
    }
    return credentials;
  } catch {
    throw new Error("Web 凭据内容无效");
  } finally {
    plaintext.fill(0);
  }
}

const selectorLookupScript = `
const findElement = (selector) => {
  const separator = selector.indexOf("=");
  const kind = selector.slice(0, separator).trim().toLowerCase();
  const value = selector.slice(separator + 1);
  if (kind === "id") return document.getElementById(value);
  if (kind === "name") return document.getElementsByName(value)[0];
  if (kind === "type") return document.querySelector('[type="' + CSS.escape(value) + '"]');
  if (kind === "class_name") return document.getElementsByClassName(value)[0];
  if (kind === "css" || kind === "css_selector") return document.querySelector(value);
  if (kind === "xpath") return document.evaluate(value, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
};`;

export function buildAutofillProbeScript(selectors) {
  return `(() => new Promise((resolve) => {
${selectorLookupScript}
const selectors = ${JSON.stringify(selectors)};
const visible = (element) => {
  if (!(element instanceof Element) || !element.isConnected || element.getClientRects().length === 0) return false;
  const style = getComputedStyle(element);
  return style.display !== "none" && style.visibility !== "hidden";
};
const elementsReady = () => {
  if (document.readyState !== "complete") return false;
  const username = selectors.username ? findElement(selectors.username) : null;
  const password = findElement(selectors.password);
  const submit = findElement(selectors.submit);
  const usernameReady = !selectors.username || (username instanceof HTMLInputElement && ["text", "email", "tel"].includes(username.type) && !username.disabled && visible(username));
  const passwordReady = password instanceof HTMLInputElement && password.type === "password" && !password.disabled && visible(password);
  const submitReady = ((submit instanceof HTMLButtonElement && !submit.disabled) || submit instanceof HTMLAnchorElement || (submit instanceof HTMLInputElement && ["submit", "button"].includes(submit.type) && !submit.disabled)) && visible(submit);
  return usernameReady && passwordReady && submitReady;
};
let readySince = 0;
let finished = false;
const finish = (ready) => {
  if (finished) return;
  finished = true;
  observer.disconnect();
  clearInterval(interval);
  clearTimeout(timeout);
  resolve(ready);
};
const check = () => {
  if (!elementsReady()) { readySince = 0; return; }
  if (!readySince) readySince = performance.now();
  if (performance.now() - readySince >= 300) finish(true);
};
const observer = new MutationObserver(check);
observer.observe(document.documentElement, { childList: true, subtree: true, attributes: true });
const interval = setInterval(check, 50);
const timeout = setTimeout(() => finish(false), 15000);
check();
}))()`;
}

export function buildAutofillScript(selectors, credentials) {
  const payload = {
    usernameValue: credentials.username,
    passwordValue: credentials.password,
    usernameSelector: selectors.username,
    passwordSelector: selectors.password,
    submitSelector: selectors.submit
  };
  return `(() => {
${selectorLookupScript}
let payload = ${JSON.stringify(payload)};
const username = payload.usernameSelector ? findElement(payload.usernameSelector) : null;
const password = findElement(payload.passwordSelector);
const submit = findElement(payload.submitSelector);
if ((payload.usernameSelector && !(username instanceof HTMLInputElement)) || !(password instanceof HTMLInputElement) || !submit) {
  payload.usernameValue = "";
  payload.passwordValue = "";
  return false;
}
const overlay = document.createElement("div");
overlay.setAttribute("data-jms-secure-login", "true");
Object.assign(overlay.style, { position: "fixed", inset: "0", zIndex: "2147483647", cursor: "wait", background: "transparent" });
document.documentElement.appendChild(overlay);
let internalAction = false;
const blockedEvents = ["pointerdown", "pointerup", "mousedown", "mouseup", "click", "dblclick", "keydown", "keyup", "keypress", "touchstart", "touchend"];
const blocker = (event) => {
  if (internalAction) return;
  event.preventDefault();
  event.stopImmediatePropagation();
};
for (const name of blockedEvents) document.addEventListener(name, blocker, true);
const setValue = (element, value) => {
  const setter = Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, "value").set;
  setter.call(element, value);
  element.dispatchEvent(new Event("input", { bubbles: true }));
  element.dispatchEvent(new Event("change", { bubbles: true }));
};
if (username) setValue(username, payload.usernameValue);
setValue(password, payload.passwordValue);
payload.usernameValue = "";
payload.passwordValue = "";
setTimeout(() => {
  for (const name of blockedEvents) document.removeEventListener(name, blocker, true);
  overlay.remove();
  if (!password.isConnected) return;
  if (username?.isConnected) setValue(username, "");
  setValue(password, "");
}, 20000);
internalAction = true;
try {
  const form = submit.closest("form");
  if (submit instanceof HTMLAnchorElement && form instanceof HTMLFormElement) {
    if (typeof form.requestSubmit === "function") form.requestSubmit();
    else form.dispatchEvent(new Event("submit", { bubbles: true, cancelable: true }));
  } else {
    submit.click();
  }
} finally {
  internalAction = false;
}
return true;
})()`;
}
